Migration trial: Update: upgrade plan screen

bogiii opened the pull request:
Closes #80174

## Proposed Changes

* Added logic to show the notification screen when a user is not eligible to try a free trial since it's already used

## Testing Instructions

* Log in as a non-a8c user (a8c user can create multiple sites with the TRIAL subscriptions)
* Go to `setup/import-focused/importerWordpress?siteSlug={SIMPLE_SITE_SLUG}&from={JN_SITE_URL }&option=everything`
* Check if "Try it for free" button has a popover message on hover/focus
<img width="805" alt="Screenshot 2023-08-14 at 13 52 41" src="https://github.com/Automattic/wp-calypso/assets/1241413/e86a07d4-e226-4965-926d-d289b2dedfff">

* Go to `setup/import-focused/migrationTrial?siteSlug={SIMPLE_SITE_SLUG}&from={JN_SITE_URL }&option=everything`
* Check if the "You already have an active free trial" screen is rendered
* Check if the "Purchase and migrate" button leads to the checkout page

**NOTE:** I would keep this screen if someone goes directly to the **migrationTrial** route

<img width="784" alt="Screenshot 2023-08-09 at 22 48 47" src="https://github.com/Automattic/wp-calypso/assets/1241413/2ca3fa80-afb1-4ef8-aa27-72358cb08617">


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

bogiii committed:
Created hook for checking is site eligible for migration trial plan

bogiii committed:
Show tooltip msg if user is not eligible to try trial plan

ouikhuan commented on client/blocks/importer/hooks/use-site-eligible-migration-trial-plan.ts:
```diff
@@ -0,0 +1,21 @@
+ * Allow only one migration trial site per user.
+ * A8c team members are allowed to have more than one.
+ */
+export function useSiteEligibleMigrationTrialPlan(): boolean {
+	const isA8c = useSelector( isA8cTeamMember );
+	const sites = useSelector( getSites );
+	const trialSites = sites.filter( ( site ) => site && isMigrationTrialSite( site ) ) || [];
+
+	if ( ! isEnabled( 'plans/migration-trial' ) ) {
+		return false;
+	}
+
+	return isA8c || ! ( trialSites.length > 0 );
+}
```
I think we probably need to have the logic in the backend and output for frontend to use it. There will be few conditions here:
- A8C users are allowed multiple migration trials
- Users with feature config `allow_multiple_migration_trials` enabled are allow multiple migration trials.
- Normal users have ever had a migration trial are not allow to create another one. This also means we will need to include the removed/inactive subscription as well. So `const trialSites = sites.filter( ( site ) => site && isMigrationTrialSite( site ) ) || [];` might not be enough. 

Maybe we can see output another property in `/me` API so we can consume this in frontend. What do you think?

bogiii replied:
It makes sense with all you said. Another property in the `/me` API sounds perfect.

ouikhuan replied:
Let's park it for now until the backend is ready, thanks @bogiii !

bogiii replied:
The `useSiteEligibleMigrationTrialPlan` function has been updated, and now we rely on the `allow_multiple_business_trials` flag.

bogiii committed:
Update useSiteEligibleMigrationTrialPlan function and rely on user flag

ouikhuan commented on client/blocks/importer/hooks/use-site-eligible-migration-trial-plan.ts:
```diff
@@ -0,0 +1,15 @@
+ */
+export function useSiteEligibleMigrationTrialPlan(): boolean {
+	const user = useSelector( getCurrentUser );
+	const flags = user?.meta?.data?.flags?.active_flags || [];
+
+	return isEnabled( 'plans/migration-trial' ) && flags.includes( 'allow_multiple_business_trials' );
+}
```
Sorry for the trouble @bogiii . I know we spoke about this and I was thinking in the wrong direction yesterday, because we miss one case here, if a user had a migration trial before on other sites then this check won't work. I've created a new endpoint to check if the trial plan is eligible or not now. See D118425-code

So you can consume it and check if `eligible` is `true` here. All other checking logic is handled in this new endpoint as well so you don't need to check the feature flags either. Please let me know if you have any questions, thank you 🙏

bogiii replied:
The custom logic has been removed. It's ready for re-review. Thanks for pointing this out!

bogiii committed:
Create check plan eligiblity query and implement to upgrade-plan comp

bogiii committed:
Remove custom logic hook for checking eligiblity

ouikhuan reviewed (approved):
LGTM 🚢  Thanks again for making the change!

javierarce commented:
I left an alternative proposal for this screen [here](https://www.figma.com/file/nRzixgCz1TsnOLPiKr4awj/Free-trials-in-the-Migrate-to-WP.com-flow?type=design&node-id=856-5582&mode=design&t=7RtcAJJLFzivjwia-4).

While it's great to explain the user what's going on right away, I think the proposal in this PR it's complex for the user, since it adds a new option (the content only import) without  context. 

I think a solution like the one I'm proposing will give us more space to explain why the trial is unavailable clearly and encourage the user to upgrade the plan.

![image](https://github.com/Automattic/wp-calypso/assets/4933/48079390-8897-4869-861d-c3a4e24baa4c)

bogiii committed:
Create notification screen

bogiii committed:
Remove tooltip

bogiii committed:
Implement checkout submission

bogiii committed:
Implement checkout submission on other flows

bogiii commented:
@ouikhuan,

I removed the previous version, and the new notification screen has been implemented based on Javier's proposal. I updated the PR description, so please re-review the PR again.

Thanks!

ouikhuan commented:
> @ouikhuan,
> 
> I removed the previous version, and the new notification screen has been implemented based on Javier's proposal. I updated the PR description, so please re-review the PR again.
> 
> Thanks!
Hmmm, seems the wording here might not fit all of the cases here. Currently the API property `eligible` with return `false` if:
- A normal user with an active migration trial subscription with current blog or other blogs under the same user account.
- A normal user with an expired/removed migration trial subscription with current blog or other blogs under the same user account.
- An A8C user (or users allow to have multiple trials) with an active migration trial subscription in the current blog and they try to add another trial plan to the same blog. (However, users in this situation won't be able to see the plans page from the frontend so we can ignore this)

To conclude, users with a migration trial in their purchase history will be marked as `eligible: false` (discussions https://github.com/Automattic/wp-calypso/issues/80173#issuecomment-1665146186). Even with the plan expired/removed, they won't be able to add another trial plan to their sites anymore. So maybe it's better for us to change the wording here to something else. @javierarce what do you think?

@bogiii, I realized the page took me to the checkout page, but it didn't redirect me back to the migration flow after I purchase the Business plan here. Since a user starts within the migration flow, should we redirect them back to it?

bogiii committed:
Handle checkout cancelation case

bogiii commented:
> I realized the page took me to the checkout page, but it didn't redirect me back to the migration flow after I purchase the Business plan here. Since a user starts within the migration flow, should we redirect them back to it?

@ouikhuan, good catch. The return and cancelation logic has been added.

It's ready for review again.

zaerl reviewed (approved):
LGTM.

bogiii commented:
No problem @javierarce, I'll revert to the previous version with the updated copy.
I need to mention that following the accessibility rules, we should not attach the tooltip to the disabled button. Even keyboard focus will not work in that case.

I'm suggesting to have a button enabled, it will be accessible and tooltip will be there.

Are you ok with that?

cc @vishnugopal

bogiii committed:
Rebase with trunk

bogiii committed:
Revert showing tooltip info for a non-eligible trial plan

bogiii commented:
Hey @javierarce,

Could you please take a look at the current implementation?
I've used the `Tooltip` component from the `@wordpress/components` package and the tooltip has dark background without the arrow.

javierarce commented:
> Could you please take a look at the current implementation?
I've used the Tooltip component from the @wordpress/components package and the tooltip has dark background without the arrow.

Isn't this one from the newsletter onboarding flow available? 
![image](https://github.com/Automattic/wp-calypso/assets/4933/aa98839c-30df-45c4-9e71-51a7160e4232)

javierarce commented:
> I'm suggesting to have a button enabled, it will be accessible and tooltip will be there.
👍

bogiii committed:
Replace Tooltip with Popover implementation

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8712897

Thank you @bogiii for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8712897).
